dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactory.java

```diff
@@ -0,0 +1,358 @@
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+
+// We want this to start before CDI so it clears the list of stopped apps before CDI gets to them.
+@Component(property = { "service.vendor=IBM",  "service.ranking:Integer=150" })
+public class OpenTelemetryInfoFactory implements ApplicationStateListener {
+
+    /*
+     * Class summary. 
+     * 
+     * This class consists of three parts
+     * 1: Static methods for producing Objects of: OpenTelemetryInfo, OpenTelemetry, Tracer, Span, Baggage.
+     *    If invoked after an app has been shutdown these will return a no-op object.
+     * 
+     * 2. A cache of OpenTelemetryInfo (using J2EENames as the key). OpenTelemetryInfo in turn contains OpenTelemetry objects. This is populated lazily using computeIfAbsent and the use of ConcurrentHashMap ensures thread safety.
+     *    
+     * 3. An ApplicationStateListener. 
+     *    On startup the ASL only removes applications from the list of apps that have shut down so they will not return a no-op
+     *    On shutdown the ASL cleans up an OpenTelemetry object, removes its parent OpenTelemetryInfo from the cache, and registers it as an app which has shut down.
+     */
+
+    private static final String INSTRUMENTATION_NAME = "io.openliberty.microprofile.telemetry";
+    private static final String ENV_DISABLE_PROPERTY = "OTEL_SDK_DISABLED";
+    private static final String CONFIG_DISABLE_PROPERTY = "otel.sdk.disabled";
+    private static final String ENV_METRICS_EXPORTER_PROPERTY = "OTEL_METRICS_EXPORTER";
+    private static final String CONFIG_METRICS_EXPORTER_PROPERTY = "otel.metrics.exporter";
+    private static final String ENV_LOGS_EXPORTER_PROPERTY = "OTEL_LOGS_EXPORTER";
+    private static final String CONFIG_LOGS_EXPORTER_PROPERTY = "otel.logs.exporter";
+    private static final String SERVICE_NAME_PROPERTY = "otel.service.name";
+
+    private static final TraceComponent tc = Tr.register(OpenTelemetryInfoFactory.class);
+
+    private static final Map<J2EEName, OpenTelemetryInfo> appToInfo = new ConcurrentHashMap<J2EEName, OpenTelemetryInfo>();
+    private static final Set<J2EEName> shutDownApplications = Collections.newSetFromMap(new WeakHashMap<J2EEName, Boolean>()); //We have no good place to say forget about an app entirely, so I use weakreferences. 
+
+    //The key methods that provide access to telemetry objects begin here
+    public static OpenTelemetryInfo getOpenTelemetryInfo(J2EEName j2EEName) {
+        return appToInfo.computeIfAbsent(j2EEName, OpenTelemetryInfoFactory::createOpenTelemetryInfo);
+    }
+
+    public static Tracer getTracer(J2EEName j2EEName) {
+        try {
+            return getOpenTelemetryInfo(j2EEName).getOpenTelemetry().getTracer(INSTRUMENTATION_NAME);
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+            return OpenTelemetry.noop().getTracerProvider().get("");
+        }
+    }
```

Comment:
Is this creating a new `Tracer` on every call?

Reply:
[Javadoc](https://www.javadoc.io/doc/io.opentelemetry/opentelemetry-api/1.29.0/io/opentelemetry/api/OpenTelemetry.html) says "gets or creates a named tracer" so it probably won't but it would be good to confirm as it could be a performance issue if we're creating a new tracer on every call.